contracts/modules/cover/Cover.sol

```diff
@@ -0,0 +1,570 @@
+      stakingPools
+    );
+
+    // make the cover expire at current block
+    uint32 previousPeriod = covers[coverId].period;
+
+    // new period is the remaining time
+    uint32 newPeriod = previousPeriod - (uint32(block.timestamp) - cover.start);
+    uint priceAlreadyPaid = (previousPeriod - newPeriod) / previousPeriod * cover.premium;
+    covers[coverId].period = newPeriod;
+
+    if (priceInAsset > priceAlreadyPaid) {
+      // get price for already paid asset
+      uint priceToBePaid = priceInAsset - priceAlreadyPaid;
+      require(priceToBePaid <= maxPrice, "Cover: Price exceeds maxPrice");
+      retrievePayment(priceToBePaid, cover.payoutAsset);
+    }
+
+    return newCoverId;
+  }
+
+  function addAmount(
+    uint coverId,
+    uint96 amount,
+    uint maxPrice,
+    StakingPool[] memory stakingPools
+  ) external returns (uint) {
+
+    require(msg.sender == ERC721.ownerOf(coverId), "Cover: not cover owner");
+
+    (uint coverId, uint premiumInAsset) = _addAmount(coverId, amount, stakingPools);
+
+    require(premiumInAsset <= maxPrice, "Cover: Price exceeds maxPrice");
+    retrievePayment(premiumInAsset, covers[coverId].payoutAsset);
+    return coverId;
+  }
+
+  function _addAmount(
+    uint coverId,
+    uint96 amount,
+    StakingPool[] memory stakingPools
+  ) internal returns (uint newCoverId, uint premiumInAsset) {
+
+    Cover memory previousCover = covers[coverId];
+    // clone the existing cover
+    Cover memory cover = covers[coverId];
+
+    StakingPool[] storage currentStakingPools = stakingPoolsForCover[covers.length];
+
+    uint32 period = uint32(block.timestamp) - cover.start;
+    // convert to NXM amount
+    uint amountToCover = uint(amount) * 1e18 / pool().getTokenPrice(pool().assets(cover.payoutAsset));
+    uint totalPremiumInNXM = 0;
+    for (uint i = 0; i < stakingPools.length; i++) {
+      if (amountToCover == 0) {
+        break;
+      }
+
+      IStakingPool stakingPool = IStakingPool(stakingPools[i].poolAddress);
+      (uint coveredAmount, uint premiumInNXM) = buyCoverFromPool(stakingPool, cover.productId, amountToCover, period);
+      amountToCover -= coveredAmount;
+      totalPremiumInNXM += premiumInNXM;
+
+      uint j = 0;
+      for ( ; j < currentStakingPools.length; j++) {
+        if (currentStakingPools[j].poolAddress == stakingPools[i].poolAddress) {
+          currentStakingPools[j].coverAmount += uint96(coveredAmount);
+          currentStakingPools[j].premiumInNXM += uint96(premiumInNXM);
+          break;
+        }
+      }
+
+      if (j < currentStakingPools.length) {
+        continue;
+      }
+
+      stakingPoolsForCover[covers.length].push(
+        StakingPool(
+          address(stakingPool),
+          uint96(coveredAmount),
+          uint96(premiumInNXM)
+        ));
+    }
+    require(amountToCover == 0, "Not enough available capacity");
+
+    premiumInAsset = totalPremiumInNXM * pool().getTokenPrice(pool().assets(cover.payoutAsset)) / 1e18;
```

Comment:
don't read the pool address and price twice. instead do this at the beginning of the function:

```
uint tokenPrice;
{
    IPool _pool = pool();
    tokenPrice = _pool.getTokenPrice(_pool.assets(originalCover.payoutAsset));
}
```